pkg/http/quirkconn.go

```diff
@@ -0,0 +1,45 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package http
+
+import (
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+// QuirkConn - similar to golang net.Conn struct, but contains a workaround of the
+// following the go bug reported here https://github.com/golang/go/issues/21133.
+// Once the bug will be fixed, we can remove this structure and replaces it with
+// the standard net.Conn
+type QuirkConn struct {
+	net.Conn
+	hadReadDeadlineInPast int32 // atomic
+}
+
+// SetReadDeadline - implements a workaround of SetReadDeadline go bug
+func (q *QuirkConn) SetReadDeadline(t time.Time) error {
```

Comment:
@harshavardhana yeah we are right, it was a mistake and I fixed it..

By the way, I really tried to make QuirkConn fully transparent.. but it wasn't possible, bufConn should be aware about the read deadline problem and should check by itself if it is allowed to set read deadline to avoid causing the go 1.8 bug.

The reason is that both golang API and bufConn calls SetReadDeadline() and we really need to synchronize those calls due to the go 1.8 bug.